### PR TITLE
typecheck trio/_dtls.py

### PR DESCRIPTION
**Pull request by jakkdl**
The remaining entries in `verify_types.json` for classes in `_dtls` is due to `_SocketType` not being fully typed yet.

### CODE REVIEW

**Comment by jakkdl** [comment]
Oops, forgot to push yesterday. But merged in `origin/master` now

File: docs/source/reference-core.rst
Comment (Zac-HD) [inline docs/source/reference-core.rst:925]: https://trio--2704.org.readthedocs.build/en/2704/reference-core.html#trio.TaskStatus doesn't seem very useful, add `:members:`?

File: docs/source/reference-io.rst
Comment (Zac-HD) [inline docs/source/reference-io.rst:307]: https://trio--2704.org.readthedocs.build/en/2704/reference-io.html#trio.DTLSChannel.statistics looks like we might want to de-duplicated these docs a bit; make the first a cross-reference to the newly-added second?

File: pyproject.toml
Comment (Zac-HD) [inline pyproject.toml:49]: I usually prefer default-strict, allowlist relaxed; but in this case I think we should continue doing it the other way for a bit longer.

Unrelated, we can delete the Pytest warnings-ignores a bit further down since we're now depending on the fixed version.

File: trio/_dtls.py
Comment (Zac-HD) [inline trio/_dtls.py:805]: I _think_ that Sphinx will do the right thing if we move this description to an attribute-docstring below, and set `:members:` in the .rst file.

File: trio/_dtls.py
Comment (Zac-HD) [inline trio/_dtls.py:1081]: quite confident that it's bytes; packets are always* sequences-of-octets.

File: trio/_dtls.py
Comment (Zac-HD) [inline trio/_dtls.py:1154]: Cross-reference newly-documented statistics object instead of re-describing here.

File: trio/_dtls.py
Comment (Zac-HD) [inline trio/_dtls.py:1278]: It looks like `*args` are positional arguments to `async_fn` which are passed after the `stream: DTLSChannel`, meaning that our type annotation for `async_fn` is incorrect to specify that it must have only one argument - and for `*args` is weaker than we could manage with `ParamSpec` and `Concatenate`.

I think we should _either_ weaken these annotations to `async_fn: Callable[..., Awaitable]`, or strengthen them to `async_fn: Callable[Concatenate[DTLSChannel, P], Awaitable], *args: P.args`.  I'd prefer the latter if it's feasible given deps and python versions.

Comment (TeamSpen210) [reply trio/_dtls.py:1278]: `ParamSpec` requires using both `*args` and `*kwargs`, only one doesn't work. For typing this we need `TypeVarTuple`, but support for that in Mypy is currently experimental and incomplete.

Comment (Zac-HD) [reply trio/_dtls.py:1278]: 😭 

Alright, let's use the simpler annotation then and leave a comment about what we want to do and that it's blocked on Mypy supporting `TypeVarTuple`.

File: pyproject.toml
Comment (jakkdl) [reply pyproject.toml:49]: yeah, was planning to switch over once the allowlist got .. not huge.

I'll open a separate PR with removing warnings.

File: trio/_dtls.py
Comment (jakkdl) [reply trio/_dtls.py:1278]: Or we could extend the API to accept `*kwargs` - though `pyright` requires all other parameters to be before both `P.args` and `P.kwargs` so that would ultimately break stuff.

Weakening the annotation and adding a comment :+1:

File: trio/_dtls.py
Comment (jakkdl) [reply trio/_dtls.py:805]: It seems that *should* work, either with `:members:` or with `:autoattribute:` but I seem to fail to find the right magical invocation to get it to work and print the attribute docstring. I've tried setting the attribute-docstring both with `#:` on the line before, or with a `"""` docstring afterwards. Anybody else willing to play around with it is very welcome to.


https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoattribute
https://stackoverflow.com/questions/22717179/sphinx-class-attribute-documentation

File: docs/source/reference-io.rst
Comment (jakkdl) [reply docs/source/reference-io.rst:307]: Should do the same with the other *Statistics.

File: trio/_dtls.py
Comment (Zac-HD) [reply trio/_dtls.py]: No kwargs; they make it impossible to backwards-compatibly add new options to the function later.  We use the allow-args/use-partial-for-kwargs style in many places.

File: docs/source/reference-core.rst
Comment (jakkdl) [reply docs/source/reference-core.rst:925]: It's ... still not very useful. The class needs a docstring and the `started` needs one too.

File: docs-requirements.in
Comment (A5rocks) [inline docs-requirements.in:24]: Maybe update the comment in `test-requirements` too?

File: trio/_dtls.py
Comment (A5rocks) [inline trio/_dtls.py:353]: I mean, also we just don't only support 3.10+ so no runtime support for this...

File: trio/_dtls.py
Comment (A5rocks) [inline trio/_dtls.py:410]: IIRC this isn't required? Is this cause mypy requires that some part of the signature be type-annotated in order to type-check the contents?

File: trio/_dtls.py
Comment (A5rocks) [inline trio/_dtls.py:567]: Is `address` `Any` for convenience's sake?

File: trio/_dtls.py
Comment (A5rocks) [inline trio/_dtls.py:743]: Huh, `weakref.ReferenceType`, I remember using `weakref.ref` as a generic. I looked at it and supposedly it's just a convenience for `ReferenceType` but it does look nicer!

File: trio/_dtls.py
Comment (A5rocks) [inline trio/_dtls.py:842]: Is this `attr-defined` for `ctx.set_options` or `SSL.X`?

File: trio/_dtls.py
Comment (A5rocks) [inline trio/_dtls.py:973]: I looked at this a couple minutes and I don't really understand why we assume `volley_messages[0]` is a `HandshakeMessage`. This might be some sort of invariant in DTLS so IDK... Probably no harm in adding an `isinstance`?

File: trio/_dtls.py
Comment (A5rocks) [inline trio/_dtls.py:1058]: Did you accidentally add a trailing comma here which made `black` want to reformat like this?

File: trio/_dtls.py
Comment (A5rocks) [inline trio/_dtls.py:1198]: Shouldn't there be generic arguments here?

File: trio/_dtls.py
Comment (A5rocks) [inline trio/_dtls.py:1260]: For what it's worth we *can* enforce that there's a `DTLSChannel` in the first positional argument via ParamSpec. It's probably not worth the trouble, though.

Umm, something like this IIRC?
```py
import typing

P = typing.ParamSpec("P")
R = typing.TypeVar("R")
Stage1 = typing.Callable[typing.Concatenate[DTLSChannel, P], typing.Awaitable[R]]
Stage2 = Stage1[..., R]
```

Then, just `async_fn: Stage2[object]`.

File: trio/_dtls.py
Comment (A5rocks) [inline trio/_dtls.py:1265]: I think this is missing a type parameter? Just `object` should work.

**Review by A5rocks — CHANGES_REQUESTED** [review]
A couple more comments!

File: trio/_dtls.py
Comment (TeamSpen210) [reply trio/_dtls.py:410]: Yeah, at least one part needs to be. Probably easiest to just always annotate.

Comment (Zac-HD) [reply trio/_dtls.py:410]: Unless we enable Mypy's check-untyped-defs option, which is usually more trouble than it's worth in this kind of migration.

File: .gitattributes
Comment (jakkdl) [reply .gitattributes:6]: I messed around with it in this PR before I wrote the small one merged into main, will be cleaned up when I merge main into this next time

File: docs-requirements.in
Comment (jakkdl) [reply docs-requirements.in:24]: it's not just used for annotations in tests, the code uses it, but it wasn't needed in docs previously until it showed up in the annotations.

File: trio/_dtls.py
Comment (jakkdl) [reply trio/_dtls.py:410]: yeah `check_untyped_defs` on the whole codebase is a long way away, but forbidding untyped function signatures will prompt to add the `-> None` where necessary on `__init__`s (as will pyright `--verifytypes`). IDR if you can require always specify `-> None` but I don't think it's much of a problem with status quo

File: trio/_dtls.py
Comment (jakkdl) [reply trio/_dtls.py:567]: It stems from the return type of `_SocketType.recvfrom`. I'll add a comment to update it in whichever gets merged later.

File: trio/_dtls.py
Comment (jakkdl) [reply trio/_dtls.py:743]: which one looks nicer? ig `ReferenceType` is a bit verbose unless you do a `from x import y` import

File: trio/_dtls.py
Comment (jakkdl) [reply trio/_dtls.py:842]: `SSL.OP_NO_RENEGOTIATION`. Coerced black into splitting across a couple lines to make it apparent which one it's referring to. No clue why it's being complained about or if it's an issue.

File: trio/_dtls.py
Comment (jakkdl) [reply trio/_dtls.py:973]: It seems to be a `HandshakeMessage` in the tests, and the only other datatype (in that file) with `msg_seq` is `HandshakeFragment`, so I suppose it should be fine.

File: trio/_dtls.py
Comment (jakkdl) [reply trio/_dtls.py:1058]: no clue! But reverted it

File: trio/_dtls.py
Comment (jakkdl) [reply trio/_dtls.py:1198]: ye, I hadn't enabled `disallow_any_generics`

File: trio/_dtls.py
Comment (jakkdl) [reply trio/_dtls.py:1260]: Making the signature incomprehensible w/o reading the context does not seem worth it, at least here, but very fancy!

File: .gitattributes
Comment (jakkdl) [reply .gitattributes:6]: Or maybe it won't, idk. Regardless, it doesn't change anything

File: docs-requirements.in
Comment (A5rocks) [reply docs-requirements.in:24]: Ah, I meant update the comment to add that it's used in annotations too.

File: trio/_dtls.py
Comment (A5rocks) [reply trio/_dtls.py:410]: Yeah I think this is fine, just was wondering cause this is different than what you've done before (... at least that I noticed)

File: trio/_dtls.py
Comment (A5rocks) [reply trio/_dtls.py]: Ah, ok.

File: trio/_dtls.py
Comment (A5rocks) [reply trio/_dtls.py]: Sorry, IMO weakref.ref looks nicer fully-qualified. But tbf this is a nit. Up to you.

File: trio/_dtls.py
Comment (A5rocks) [reply trio/_dtls.py:1264]: Yup. Agreed. FWIW there's an open mypy issue about loosening the syntax to not require type aliasing, but I don't remember there being consensus about it.

File: trio/_dtls.py
Comment (A5rocks) [inline trio/_dtls.py:353]: FWIW I don't think this comment is necessary, as long as we remember to run `pyupgrade` (or manually ctrl+f for unions) we'll manage. I guess this is up to you though.

File: trio/_dtls.py
Comment (A5rocks) [inline trio/_dtls.py:648]: TBF should this be `_T` since (unlike other files, where this is just implied from the filename... IDK why this is different.) we're doing the whole "underscore for private" convention.

**Review by A5rocks — APPROVED** [review]
This seems fine. Just a couple minor nitpicks. Feel free to merge! (Maybe wait a day to see if anyone chimes in?)

File: trio/_dtls.py
Comment (A5rocks) [reply trio/_dtls.py:353]: There's another comment like this in another file (forgot its name and forgot to comment when I saw it), so above applies to that too.

Comment (jakkdl) [reply trio/_dtls.py:353]: that one spawned enough confusion (mostly in me probably) that I ended up writing a proper comment, so I don't think leaving that one hurts.

**Comment by jakkdl** [comment]
Eh, they've had two weeks to chime in and it's blocking moar typing PRs - so I'll merge.

**Pull request merged**
